util/predicates/cluster_predicates.go

```diff
@@ -59,3 +63,84 @@ func processIfClusterNotImported(logger logr.Logger, obj client.Object) bool {
+		},
+		GenericFunc: func(e event.GenericEvent) bool {
+			return processIfClusterReadyControlPlane(logger.WithValues("predicate", "ClusterWithReadyControlPlane", "eventType", "generic"), e.Object)
+		},
+	}
+}
+
+func processIfClusterReadyControlPlane(logger logr.Logger, obj client.Object) bool {
+	kind := strings.ToLower(obj.GetObjectKind().GroupVersionKind().Kind)
+	log := logger.WithValues("namespace", obj.GetNamespace(), kind, obj.GetName())
+
+	cluster, ok := obj.(*clusterv1.Cluster)
+	if !ok {
+		log.V(4).Info("Expected a Cluster but got a different object, will not attempt to map resource", "object", obj)
+		return false
+	}
+
+	if !cluster.Status.ControlPlaneReady {
+		log.V(4).Info("Cluster does not have an ready control plane, will not attempt to map resource")
+		return false
+	}
+
+	log.V(6).Info("Cluster has an ready control plane, will attempt to map resource")
+
+	return true
+}
+
+// ClusterOrNamespaceWithImportLabel returns a predicate that returns true only if the provided resource is a cluster and
+// has an import label set on it or on its namespace.
+func ClusterOrNamespaceWithImportLabel(ctx context.Context, logger logr.Logger, cl client.Client, label string) predicate.Funcs {
+	return predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return processIfClusterOrNamespaceWithImportLabel(ctx,
+				logger.WithValues("predicate", "ClusterOrNamespaceWithImportLabel", "eventType", "update"), cl, e.ObjectNew, label)
+		},
+		CreateFunc: func(e event.CreateEvent) bool {
+			return processIfClusterOrNamespaceWithImportLabel(ctx,
+				logger.WithValues("predicate", "ClusterOrNamespaceWithImportLabel", "eventType", "create"), cl, e.Object, label)
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return processIfClusterOrNamespaceWithImportLabel(ctx,
+				logger.WithValues("predicate", "ClusterOrNamespaceWithImportLabel", "eventType", "delete"), cl, e.Object, label)
+		},
+		GenericFunc: func(e event.GenericEvent) bool {
+			return processIfClusterOrNamespaceWithImportLabel(ctx,
+				logger.WithValues("predicate", "ClusterOrNamespaceWithImportLabel", "eventType", "generic"), cl, e.Object, label)
+		},
+	}
+}
+
+func processIfClusterOrNamespaceWithImportLabel(ctx context.Context, logger logr.Logger, cl client.Client, obj client.Object, label string) bool {
+	kind := strings.ToLower(obj.GetObjectKind().GroupVersionKind().Kind)
+	log := logger.WithValues("namespace", obj.GetNamespace(), kind, obj.GetName())
+
+	cluster, ok := obj.(*clusterv1.Cluster)
+	if !ok {
+		log.V(4).Info("Expected a Cluster but got a different object, will not attempt to map resource", "object", obj)
+		return false
+	}
+
+	shouldImport, err := util.ShouldAutoImport(ctx, log, cl, cluster, label)
+	if err != nil {
+		log.Error(err, "Client failed to get namespace, this error is ignored returning true")
```

Comment:
this is not always the case, `ShouldAutoImport` can return true also when there is an import annotation set on the cluster object. Reading this log does not help much tbh, I would suggest rather having something like:

```suggestion
		log.Error(err, "namespace or cluster has already import annotation set, ignoring it")
```